Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
 #else
     UNUSED_PARAM(didWrite);
 #endif
+}
+
+void NetworkStorageManager::spaceGrantedForOrigin(const WebCore::ClientOrigin& origin, uint64_t spaceRequested)
+{
+    assertIsCurrent(workQueue());
+
+    updateLastModificationTimeForOrigin(origin);
+    if (!m_totalQuotaRatio)
+        return;
+
+    if (!m_totalQuota) {
+        std::optional<uint64_t> volumeCapacity;
+        if (m_volumeCapacityOverride)
+            volumeCapacity = m_volumeCapacityOverride;
+        else if (auto capacity = FileSystem::volumeCapacity(m_path))
+            volumeCapacity = WTF::roundUpToMultipleOf(defaultVolumeCapacityUnit, *capacity);
+        if (volumeCapacity)
+            m_totalQuota = *m_totalQuotaRatio * *volumeCapacity;
+        else
+            return;
+    }
 
-    if (m_quotaManager)
-        m_quotaManager->originVisited(origin);
+    if (m_totalUsage)
+        m_totalUsage = *m_totalUsage + spaceRequested;
+
+    if (!m_totalUsage || *m_totalUsage > *m_totalQuota)
+        schedulePerformEviction();
 }
 
-NetworkQuotaManager& NetworkStorageManager::quotaManager()
+void NetworkStorageManager::schedulePerformEviction()
 {
-    if (!m_quotaManager) {
-        Vector<std::pair<WebCore::ClientOrigin, WallTime>> origins;
-        forEachOriginDirectory([&](auto directory) {
-            auto originFile = originFilePath(directory);
-            auto modificationTime = FileSystem::fileModificationTime(originFile);
-            if (modificationTime) {
-                if (auto origin = readOriginFromFile(originFile))
-                    origins.append({ *origin, *modificationTime });
-            }
-        });
+    assertIsCurrent(workQueue());
+
+    if (m_isEvictionScheduled)
+        return;
+
+    m_isEvictionScheduled = true;
+    getDomainsWithLastAccessedTimeForEviction();
+}
+
+void NetworkStorageManager::getDomainsWithLastAccessedTimeForEviction()
+{
+    assertIsCurrent(workQueue());
+
+    RunLoop::main().dispatch([this, weakThis = ThreadSafeWeakPtr { *this }]() {
+        auto strongThis = weakThis.get();
+        if (!strongThis || m_closed || !m_process)
+            return;
 
-        std::sort(origins.begin(), origins.end(), [&](const auto& a, const auto& b) {
-            return a.second < b.second;
+        m_process->registrableDomainsWithLastAccessedTime(m_sessionID, [this, weakThis = ThreadSafeWeakPtr { *this }](auto result) mutable {
+            auto strongThis = weakThis.get();
+            if (!strongThis || m_closed)
+                return;
+
+            m_queue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, result = crossThreadCopy(result)]() mutable {
+                if (auto strongThis = weakThis.get())
```

Comment:
`NetworkStorageManager` will be destroyed in the main thread because `NetworkStorageManager::close()`, which sets `m_closed` should dispatch a last task (the last task dispatches a last task reply to main thread), and we have assert in `~NetworkStorageManager`.

Note this function does not dispatch task to m_queue when `m_closed` is true.

Reply:
Then, why should we even have a weakThis/strongThis approach?
It seems we could just capture 'this', maybe add a postTask so that we understand this is fine to do like in other parts of the code base.

Reply:
This is good point; I will file a bug to make other places to only capture `this`.